src/crimson/net/SocketConnection.cc

```diff
@@ -65,38 +64,100 @@ SocketConnection::get_messenger() const {
 seastar::future<bool> SocketConnection::is_connected()
 {
   return seastar::smp::submit_to(shard_id(), [this] {
-      return !send_ready.failed();
+      return write_state == write_state_t::open;
     });
 }
 
-seastar::future<> SocketConnection::send(MessageRef msg)
+void SocketConnection::write_event()
 {
-  return seastar::smp::submit_to(shard_id(), [this, msg=std::move(msg)] {
-      if (state == state_t::closing)
-        return seastar::now();
-      return seastar::with_gate(pending_dispatch, [this, msg=std::move(msg)] {
-          return do_send(std::move(msg))
-            .handle_exception([this] (std::exception_ptr eptr) {
-              logger().warn("{} send fault: {}", *this, eptr);
+  if (write_dispatching) {
+    // already dispatching
+    return;
+  }
+  write_dispatching = true;
+  switch (write_state) {
+   case write_state_t::open:
+   case write_state_t::delay:
+    seastar::with_gate(pending_dispatch, [this] {
+      return seastar::repeat([this] {
+        switch (write_state) {
+         case write_state_t::open:
+          return seastar::futurize_apply([this] {
+              if (_keepalive) {
+                return do_keepalive()
+                  .then([this] { _keepalive = false; });
+              }
+              return seastar::now();
+            }).then([this] {
+              if (_keepalive_ack) {
+                return do_keepalive_ack()
+                  .then([this] { _keepalive_ack = false; });
+              }
+              return seastar::now();
+            }).then([this] {
+              if (!out_q.empty()){
+                MessageRef msg = out_q.front();
+                return write_message(msg)
+                  .then([this, msg] {
```

Comment:
Could you be more specific?

Reply:
the `write_event()` is growing into a lengthy function. so i guess it would be better to avoid the newlines before `.then()` if the line length does not breach 80 chars limit or it hurts the readability.

Reply:
done